File: backend/app/schemas/sanpham/SanPham.py
# /backend/app/schemas/SanPham.py
from pydantic import BaseModel, Field, ConfigDict, validator
from typing import List, Optional, Any, Dict
from datetime import datetime
from decimal import Decimal

from ..Shared import TrangThaiSanPhamEnum
from .DanhMuc import DanhMucResponse
from .CapDo import CapDoResponse
from .ThuongHieu import ThuongHieuResponse
from .BienTheSanPham import BienTheSanPhamCreate, BienTheSanPhamResponse, BienTheSanPhamUpdate


class SanPhamBase(BaseModel):
    ten_san_pham: str = Field(..., max_length=100, description="Tên sản phẩm")
    mo_ta: Optional[str] = Field(None, description="Mô tả sản phẩm")
    thong_so_ky_thuat: Optional[Dict[str, Any]] = Field(None, description="Thông số kỹ thuật")
    ngay_tao: Optional[datetime] = Field(None, description="Ngày tạo sản phẩm")
    ngay_cap_nhat: Optional[datetime] = Field(None, description="Ngày cập nhật sản phẩm")

    model_config = ConfigDict(from_attributes=True)


class SanPhamCreate(SanPhamBase):
    danh_muc_id: int = Field(..., description="ID của danh mục sản phẩm")
    thuong_hieu_id: int = Field(..., description="ID của thương hiệu sản phẩm")
    cap_do_id: int = Field(..., description="ID của cấp độ sản phẩm")
    cac_bien_the: Optional[List[BienTheSanPhamCreate]] = Field(  # ĐỔI TÊN: từ bien_the_san_phams
        [], description="Danh sách biến thể sản phẩm"
    )


class SanPhamUpdate(BaseModel):
    danh_muc_id: Optional[int] = Field(None)
    thuong_hieu_id: Optional[int] = Field(None)
    cap_do_id: Optional[int] = Field(None)
    ten_san_pham: Optional[str] = Field(None, max_length=100)
    mo_ta: Optional[str] = Field(None)
    thong_so_ky_thuat: Optional[Dict[str, Any]] = Field(None)
    cac_bien_the: Optional[List[BienTheSanPhamUpdate]] = Field(None, description="Danh sách biến thể cập nhật")
    bien_the_xoa_ids: Optional[List[int]] = Field([], description="IDs biến thể cần xóa")

    model_config = ConfigDict(from_attributes=True)


class SanPhamDelete(BaseModel):
    id: int


class SanPhamResponse(SanPhamBase):
    id: int
    ma_san_pham: str = Field(..., max_length=24, description="Mã sản phẩm")
    danh_muc: DanhMucResponse
    cap_do: CapDoResponse
    thuong_hieu: ThuongHieuResponse
    so_sao_trung_binh: Optional[Decimal] = Field(0, description="Điểm trung bình")
    so_luong_danh_gia: Optional[int] = Field(0, description="Tổng số đánh giá")
    cac_bien_the: List[BienTheSanPhamResponse] = Field(
        default_factory=list,
        description="Danh sách biến thể (đã có ảnh, giá, tồn kho)"
    )
    model_config = ConfigDict(from_attributes=True)


class SanPhamPublic(SanPhamResponse):
    trung_binh_danh_gia: Optional[Decimal] = Field(None, description="Điểm trung bình")
    so_luong_danh_gia: Optional[int] = Field(0, description="Tổng số đánh giá")


class SanPhamListResponse(BaseModel):
    data: List[SanPhamResponse] = Field(..., description="Danh sách sản phẩm")
    pagination: Dict[str, int] = Field(..., description="page, per_page, total, pages")

    model_config = ConfigDict(from_attributes=True)


class SanPhamBasicResponse(BaseModel):
    """
    Schema trả về thông tin cơ bản của sản phẩm (chỉ id và tên)
    """
    id: int
    ten_san_pham: str = Field(..., description="Tên sản phẩm")
    
    model_config = ConfigDict(from_attributes=True)


class SanPhamBasicListResponse(BaseModel):
    """
    Schema trả về danh sách sản phẩm cơ bản
    """
    data: List[SanPhamBasicResponse] = Field(..., description="Danh sách sản phẩm cơ bản")
    
    model_config = ConfigDict(from_attributes=True)

class SanPhamLienQuanResponse(SanPhamBase):
    id: int
    ma_san_pham: str = Field(..., max_length=24, description="Mã sản phẩm")
    # danh_muc: DanhMucResponse
    # cap_do: CapDoResponse
    # thuong_hieu: ThuongHieuResponse
    so_sao_trung_binh: Optional[Decimal] = Field(0, description="Điểm trung bình")
    so_luong_danh_gia: Optional[int] = Field(0, description="Tổng số đánh giá")
    cac_bien_the: List[BienTheSanPhamResponse] = Field(
        default_factory=list,
        description="Danh sách biến thể (đã có ảnh, giá, tồn kho)"
    )
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/validation/san_pham.js
// validation/san_pham.js
import * as yup from "yup";

import { TrangThaiSanPhamEnum } from "./shared"; // Giả sử export bằng ES Module
import { danh_muc_response_schema } from "./danh_muc";
import { thuong_hieu_response_schema } from "./thuong_hieu";
import {
  bien_the_san_pham_create_schema,
  bien_the_san_pham_response_schema,
  bien_the_san_pham_update_schema,
} from "./bien_the_san_pham";

const san_pham_base_schema = yup.object({
  danh_muc_id: yup.number().required("Danh mục ID bắt buộc"),
  thuong_hieu_id: yup.number().required("Thương hiệu ID bắt buộc"),
  ten_san_pham: yup.string().max(200).required("Tên sản phẩm bắt buộc"),
  mo_ta: yup.string().nullable(),
  thong_so_ky_thuat: yup.object().nullable(),
  trang_thai: yup
    .mixed()
    .oneOf(Object.values(TrangThaiSanPhamEnum))
    .default(TrangThaiSanPhamEnum.DANG_BAN),
  ngay_tao: yup.date().nullable(),
  ngay_cap_nhat: yup.date().nullable(),
});

export const san_pham_create_schema = san_pham_base_schema.shape({
  bien_the_san_phams: yup
    .array()
    .of(bien_the_san_pham_create_schema)
    .default([]),
});

export const san_pham_update_schema = yup.object({
  danh_muc_id: yup.number().nullable(),
  thuong_hieu_id: yup.number().nullable(),
  ten_san_pham: yup.string().max(200).nullable(),
  mo_ta: yup.string().nullable(),
  thong_so_ky_thuat: yup.object().nullable(),
  trang_thai: yup.mixed().oneOf(Object.values(TrangThaiSanPhamEnum)).nullable(),
  cac_bien_the: yup.array().of(bien_the_san_pham_update_schema).nullable(),
});

export const san_pham_delete_schema = yup.object({
  id: yup.number().required("ID bắt buộc"),
});

export const san_pham_response_schema = san_pham_base_schema.shape({
  id: yup.number().required(),
  ma_san_pham: yup.string().max(24).required(),
  danh_muc: danh_muc_response_schema.required(),
  thuong_hieu: thuong_hieu_response_schema.required(),
  cac_bien_the: yup.array().of(bien_the_san_pham_response_schema).default([]),
});

export const san_pham_public_schema = san_pham_response_schema.shape({
  trung_binh_danh_gia: yup.number().nullable(),
  so_luong_danh_gia: yup.number().default(0),
});

export const san_pham_list_response_schema = yup.object({
  data: yup.array().of(san_pham_response_schema).required(),
  pagination: yup
    .object({
      page: yup.number().required(),
      per_page: yup.number().required(),
      total: yup.number().required(),
      pages: yup.number().required(),
    })
    .required(),
});
